### PR TITLE
API / Web Namespace Configuration

### PR DESCRIPTION
**Pull request by ben-toogood**
Depends on https://github.com/micro/go-micro/pull/1496

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -343,10 +352,15 @@ func Commands(options ...micro.Option) []*cli.Command {
...
-				Name:    "namespace",
-				Usage:   "Set the namespace used by the API e.g. com.example.api",
+				Name:    "api_namespace",
+				Usage:   "Set the service namespace used by the API e.g. com.example.api",
 				EnvVars: []string{"MICRO_API_NAMESPACE"},
 			},
+			&cli.StringFlag{
+				Name:    "namespace",
+				Usage:   "Set the namespace used by the API e.g. micro",
+				EnvVars: []string{"MICRO_NAMESPACE"},
```
Comment (asim) [inline api/api.go:362]: This should be added as a global flag rather than in the api.

File: api/api.go [hunk 2]
```diff
@@ -46,7 +47,8 @@ var (
 	RPCPath               = "/rpc"
 	APIPath               = "/"
 	ProxyPath             = "/{service:[a-zA-Z0-9]+}"
-	Namespace             = "go.micro.api"
+	Namespace             = auth.DefaultNamespace
```
Comment (asim) [inline api/api.go:50]: This should just be referenced from the global context

File: api/api.go [hunk 3]
```diff
@@ -46,7 +47,8 @@ var (
...
+	APINamespace          = "go.micro.api"
```
Comment (asim) [inline api/api.go:51]: This should remain Namespace as its scoped as the api.Namespace

File: api/api.go [hunk 4]
```diff
@@ -343,10 +352,15 @@ func Commands(options ...micro.Option) []*cli.Command {
 				EnvVars: []string{"MICRO_API_HANDLER"},
 			},
 			&cli.StringFlag{
-				Name:    "namespace",
-				Usage:   "Set the namespace used by the API e.g. com.example.api",
+				Name:    "api_namespace",
```
Comment (asim) [inline api/api.go:355]: Should remain namespace since its scoped as micro api --namespace

File: web/web.go [hunk 5]
```diff
@@ -48,11 +48,13 @@ var (
 	Name = "go.micro.web"
 	// Default address to bind to
 	Address = ":8082"
-	// The namespace to serve
+	// The web namespace to serve
 	// Example:
 	// Namespace + /[Service]/foo/bar
 	// Host: Namespace.Service Endpoint: /foo/bar
-	Namespace = "go.micro.web"
+	WebNamespace = "go.micro.web"
```
Comment (asim) [inline web/web.go:55]: Same as api

File: cmd/cmd.go [hunk 6]
```diff
@@ -171,12 +171,6 @@ func setup(app *ccli.App) {
 			EnvVars: []string{"MICRO_REPORT_USAGE"},
 			Value:   true,
 		},
-		&ccli.StringFlag{
```
Comment (asim) [inline cmd/cmd.go:174]: Would it not make sense to have this considering all core services are probably going to depend on namespace?

File: internal/namespace/resolver/resolver.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+	"net/http"
+	"strings"
+
+	"github.com/micro/go-micro/v2/api/resolver"
+	"github.com/micro/go-micro/v2/auth"
+	"github.com/micro/go-micro/v2/logger"
+	"golang.org/x/net/publicsuffix"
+)
+
+func NewNamespaceResolver(srvType, namespace string) resolver.NamespaceResolver {
```
Comment (asim) [inline internal/namespace/resolver/resolver.go:14]: NewResolver

File: web/web.go [hunk 8]
```diff
@@ -52,7 +53,9 @@ var (
 	// Example:
 	// Namespace + /[Service]/foo/bar
 	// Host: Namespace.Service Endpoint: /foo/bar
-	Namespace = "go.micro.web"
+	Namespace     = "go.micro"
+	Type          = "web"
+	FullNamespace = "go.micro.web"
```
Comment (asim) [inline web/web.go:58]: Would just remove this since you have Namespace + Type

File: cmd/cmd.go (hunk 6 above)
Comment (ben-toogood) [reply cmd/cmd.go:174]: All core services will use determine the namepsace on a request by request basis.

File: internal/namespace/resolver/resolver.go (hunk 7 above)
Comment (ben-toogood) [reply internal/namespace/resolver/resolver.go]: Shall amend

File: web/web.go (hunk 8 above)
Comment (ben-toogood) [reply web/web.go:59]: Sure, i'll amend

**Pull request merged**